### DISCUSSION TITLE
Raycasting

### DISCUSSION THREAD

**Discussion by ThomasGorisse** [google/filament#3827] [Q&A]
Hi Filament Team,

I'm currently working on the Collision part of an upcoming ARCore/Filament repo.

Since there is not yet any raycasting implementation available on Filament from what I saw, I have 2 choices:
1. Use the ARCore `hitTest()` from [Frame](https://developers.google.com/ar/reference/java/com/google/ar/core/Trackable) based on the `Scene` [Anchors](https://developers.google.com/ar/reference/java/com/google/ar/core/Anchor) `Entity`s which is quite high level since it can take a `MotionEvent` or `Ray` parameter and return the corresponding real world [Pose](https://developers.google.com/ar/reference/java/com/google/ar/core/Pose) that I can use to detect collision with the `FilamentAsset` bounding box.
**BUT this involve excluding the non AR (Filament only) usage in the framework.**
2. Implement the Collision system from scratch inside the framework and use the ARCore HitTest when the renderable is anchored in the real world.

My questions:
1. Is there any plan to implement raycasting/collision directly inside Filament ?
2. Do you have any link/infos on already implemented raycasting/collision systems ?

Thanks

**Comment by romainguy** [comment]
We don't have plans to add a collision system. In the past @pixelflinger and I have used the Bullet physics engine for this, with great results.

**Reply by jnmaloney** [reply]
Is there a shortcut in Filament to get a ray from the camera?

**Reply by romainguy** [reply]
How so? The camera position is all you need.

**Reply by jnmaloney** [reply]
I mean a ray for mouse picking, eg giving a direction vector based on camera view, projection and relative mouse position. 

Edit: At the moment I'm doing something like this to get the mouse ray:
```
  float near_plane_distance = fila_cam->getNear();
  filament::math::float3 forward = fila_cam->getForwardVector();
  filament::math::float3 up = fila_cam->getUpVector();
  filament::math::float3 left = fila_cam->getLeftVector();
  filament::math::float3 ray_dir = relative_mouse_x * left + relative_mouse_y * up + near_plane_distance * forward;
```

**Reply by pixelflinger** [reply]
You want to take your position in clip space and apply the inverse of the projection matrix, which will give you a ray in view space, which you need to normalize.

If you need world space, you need to  also apply the inverse transpose of the model matrix.

To go from mouse coords to clip space, it's just a remapping of x and y between -1 and 1. And z doesn't matter, so just use 0.
Also don't forget that clip space (-1,-1) is left-bottom (not left-top).

**Comment by xumo** [comment]
Sorry for insisting on this, but I really need to support 3d picking on GLB items.
Now I'm stuck because in the example of the ModelViewer the AABB are returning zero center with zero extend:
```kotlin
for ( ent in modelViewer.asset!!.entities){
    if( modelViewer.engine!!.renderableManager.hasComponent(ent)){
        val  box: Box = Box();
        modelViewer.engine!!.transformManager.getInstance(ent)
        Log.v("ModelLoader",  "Box (${box.center[0]} ${box.center[1]} ${box.center[2]})" +
                        " (${box.halfExtent[0]} ${box.halfExtent[1]} ${box.halfExtent[2]})")

    }
}
```
this ouputs something like: `Box (0.0 0.0 0.0) (0.0 0.0 0.0)`

I've seen @romainguy recommending using Bullet, but I don't see how or where to start integrating it in Android: should it be at C++ level? How or where do I pass the primitives to Bullet?
That said, is there a ay I can have access to the geometries of a GLTF?

**Reply by romainguy** [reply]
@xumo You are printing the box you just created. You need to get the bounding box from the renderable manager by calling `renderableManager.getAxisAlignedBoundingBox(instance)`.

**Reply by xumo** [reply]
Sorry about that, its a little late  here
This was missing ` modelViewer.engine!!.renderableManager.getAxisAlignedBoundingBox(ent,box);`

```kotlin
for ( ent in modelViewer.asset!!.entities){
    if( modelViewer.engine!!.renderableManager.hasComponent(ent)){
        val  box: Box = Box();
        modelViewer.engine!!.renderableManager.getAxisAlignedBoundingBox(ent,box);
        Log.v("ModelLoader",  "Box (${box.center[0]} ${box.center[1]} ${box.center[2]})" +
                        " (${box.halfExtent[0]} ${box.halfExtent[1]} ${box.halfExtent[2]})")

    }
}
```
I´m using  an entity given by `modelViewer.asset!!.entities`, is that different from an instance?

**Reply by romainguy** [reply]
Yes, you first need to retrieve the component instance from the `RenderableManager` using `getInstance(entity)`.

**Reply by xumo** [reply]
OK, that works. But the AABB its center in zero, to get its current position I nee to apply the word transformation and the the model matrix?

**Comment by miinso** [comment]
@xumo I've implemented a rudimentary physics simulation [demo](https://github.com/61315/filament-physics) using Bullet SDK and Filament engine. This demo might get you started. See the References section for the details and other implementation cases.

https://github.com/google/filament/discussions/3575#discussioncomment-959590
> @paulorrfilho There is a [Java port](https://pybullet.org/Bullet/phpBB3/viewtopic.php?t=1827) of Bullet physics SDK that can be integrated to Android runtime without the burden.
> 
> I tried to glue them together and it did kind of work, still I'm positive using JNI binding will earn you some extra performance though.
> 
> [![filament-physics-cube-instancing](https://user-images.githubusercontent.com/46559594/124067157-dcc6a700-da74-11eb-8b40-715902b62f79.gif)](https://github.com/61315/filament-physics)
> 
> In terms of gravity, the Java port implements the original methods as-is, see [src/main/java/com/bulletphysics/dynamics/DiscreteDynamicsWorld.java#L409-L419](https://github.com/bubblecloud/jbullet/blob/00b51569461391d87436369ac723e875efef0dcd/src/main/java/com/bulletphysics/dynamics/DiscreteDynamicsWorld.java#L409-L419):
> 
> ```java
> @Override
> public void setGravity(Vector3f gravity) {
> 	this.gravity.set(gravity);
> 	for (int i = 0; i < collisionObjects.size(); i++) {
> 		CollisionObject colObj = collisionObjects.getQuick(i);
> 		RigidBody body = RigidBody.upcast(colObj);
> 		if (body != null) {
> 			body.setGravity(gravity);
> 		}
> 	}
> }
> ```

**Comment by romainguy** [comment]
Picking is now available in Filament using the `pick()` method on the `View` class. A picking result will return the `Entity` that was picked as well as extra information like the depth of the pixel used for the picking query.

**Reply by ThomasGorisse** [reply]
Awesome!!!

**Reply by flykule** [reply]
If you are working with dragging/animationing model, then this  pick will not work properly unfortunatelly...

**Reply by romainguy** [reply]
Why not?

**Reply by flykule** [reply]
I hava a gltf model car. It will return correct entity id in init position. After some camera translation and scaling, the entity id return by pick(with position of screen touch event) is likely to be something else. Usually it is not the closet entity beneath user's click location.

**Reply by romainguy** [reply]
I still don't understand? The pick API will return the first entity as the specified screen space location, why is that a problem? What would you like exactly?

**Reply by flykule** [reply]
Yes it will return correct entity when no dragging, but it will not if you drag your model a little. Let's say your gltf has animation that will change it's entity location after animation start, it will be very hard to picking the animationing entities. Can we implement something like three js's raycaster demo? (https://threejs.org/examples/#webgl_instancing_raycast)

**Reply by romainguy** [reply]
Why would dragging the model change anything? You can just not redo the picking for the duration of the drag. As for animation, it sounds like you don’t want pixel accurate picking then? If so you can just raycast against the bounding boxes of your entities. And you could implement that three.js demo with our picking API.

**Reply by flykule** [reply]
Okay, may be i should choose the second way and give it a try.

**Reply by zhangbo-0213** [reply]
when I use pick() function, the id of renderable in result I got is always the same value, whichever I click the renderable in scene, is there something wrong that I called function?

**Comment by ThomasGorisse** [comment]
Is there still any way to pick a semi transparent renderable?
I understand that's it's pixel picking so we might use another collision system if we want to get a ray cast to a bounding box instead of the model rendered parts but it's also frequent to have an opacity on some materials.
